### PR TITLE
8310577: Fix -Wconversion warnings in interpreter and oops

### PR DESCRIPTION
**Pull request by coleenp**
Please review this change that's mostly casts and a few int type narrowing changes.  The casts are for size calculations between two addresses within the same area so can be int and are never going to be 64 bit.
I didn't change CompileThreshold to be 'int' because SA and JVMCI code expects it to be intx.  Maybe that should change though.
Tested with tier1 on Oracle platforms, and tier1-4 linux and windows x64.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8310577](https://bugs.openjdk.org/browse/JDK-8310577): Fix -Wconversion warnings in interpreter and oops (**Enhancement** - P4)

### Reviewers
 * [Frederic Parain](https://openjdk.org/census#fparain) (@fparain - **Reviewer**)
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14604/head:pull/14604` \
`$ git checkout pull/14604`

Update a local copy of the PR: \
`$ git checkout pull/14604` \
`$ git pull https://git.openjdk.org/jdk.git pull/14604/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14604`

View PR using the GUI difftool: \
`$ git pr show -t 14604`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14604.diff">https://git.openjdk.org/jdk/pull/14604.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14604#issuecomment-1601594087)

### CODE REVIEW

File: src/hotspot/share/interpreter/invocationCounter.hpp [hunk 1]
```diff
@@ -68,7 +68,7 @@ class InvocationCounter {
   // Accessors
   bool carry() const               { return (_counter & carry_mask) != 0; }
   uint count() const               { return _counter >> number_of_noncount_bits; }
-  uint limit() const               { return CompileThreshold; }
+  uint limit() const               { return (uint)CompileThreshold; }
```
Comment (fparain) [inline src/hotspot/share/interpreter/invocationCounter.hpp:71]: CompileThreshold is an intx, and can be set by users on the command line. Casting it to uint looks wrong. This method seems to be used in only one place (invocationCounter.cpp:69), so changing the return type of the method to intx and fixing the print statement where it is used could be a better fix.

**Review by fparain — CHANGES_REQUESTED** [review]
Overall, looks good. One fix could be better in invocationCounter.hpp.

File: src/hotspot/share/interpreter/invocationCounter.hpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/interpreter/invocationCounter.hpp]: That was a good suggestion.  Thank you!

**Review by fparain — APPROVED** [review]

File: src/hotspot/share/interpreter/bytecodeUtils.cpp [hunk 2]
```diff
@@ -353,7 +353,7 @@ static void print_local_var(outputStream *os, unsigned int bci, Method* method,
 
 StackSlotAnalysisData::StackSlotAnalysisData(BasicType type) : _bci(INVALID), _type(type) {}
 
-StackSlotAnalysisData::StackSlotAnalysisData(int bci, BasicType type) : _bci(bci), _type(type) {
+StackSlotAnalysisData::StackSlotAnalysisData(int bci, BasicType type) : _bci((u2)bci), _type(type) {
```
Comment (matias9927) [inline src/hotspot/share/interpreter/bytecodeUtils.cpp:356]: Looking at this more closely, it seems like bci is meant to be a u2 but it needs one extra bit for the "INVALID" value. Maybe it's possible to represent invalid bci's differently? That way we can just set _bci to be a u2.

File: src/hotspot/share/interpreter/bytecodes.cpp [hunk 3]
```diff
@@ -439,7 +439,7 @@ void Bytecodes::def_flags(Code code, const char* format, const char* wide_format
   int len  = (format      != nullptr ? (int) strlen(format)      : 0);
   int wlen = (wide_format != nullptr ? (int) strlen(wide_format) : 0);
 #endif
-  int bc_flags = 0;
+  jchar bc_flags = 0;
   if (can_trap)           bc_flags |= _bc_can_trap;
```
Comment (matias9927) [inline src/hotspot/share/interpreter/bytecodes.cpp:443]: I believe that bitwise-or will promote the operands to int. Will this cause any issues here?

File: src/hotspot/share/interpreter/bytecodeUtils.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/interpreter/bytecodeUtils.cpp:356]: If I pass u2 bci, it'll fit in the field.  The INVALID bit is added after bci - ie the bci is never INVALID.

**Comment by coleenp** [comment]
Thanks Fred and Matias.
/integrate

**Comment by RealLucy** [comment]
Breaks the build, unfortunately. See JDK-8310735.

**Pull request closed without merging**